### PR TITLE
Avoid repeated DB queries when external claims are retrieving from DB.

### LINKED ISSUE #2990: Repeated DB calls when external claims are retrieved
**Description:**

When we are retrieving external claims from DB, we could see several SQL queries are happening repeatedly with a different value. These DB transactions induce an overhead. Hence it is better to optimize the process.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### PR DESCRIPTION
**Pull request by dinusha92**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]
Fixes #2990 
-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAO.java [hunk 1]
```diff
@@ -259,10 +243,61 @@ private String getClaimMapping(Connection connection, int externalClaimId, int t
...
+        } catch (SQLException e) {
+            throw new ClaimMetadataException("Error while listing claims for dialect " + claimDialectURI, e);
+        } finally {
+            IdentityDatabaseUtil.closeResultSet(rs);
+            IdentityDatabaseUtil.closeStatement(prepStmt);
+        }
+        return new ArrayList<ExternalClaim>(claimMap.values());
+    }
+
+
```
Comment (tharindu-bandara) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAO.java:300]:
```suggestion

```

**Comment by tharakawijekoon** [comment]
During server startup following error is observed when using an Oracle DB.

```Caused by: Error : 933, Position : 133, Sql = SELECT CLAIMS.ID, PROPERTY.PROPERTY_NAME, PROPERTY.PROPERTY_VALUE, CLAIMS.CLAIM_URI, RESOLVED.CLAIM_URI AS MAPPED_URI FROM IDN_CLAIM AS CLAIMS LEFT JOIN IDN_CLAIM_PROPERTY AS PROPERTY ON PROPERTY.LOCAL_CLAIM_ID=CLAIMS.ID AND CLAIMS.TENANT_ID=PROPERTY.TENANT_ID INNER JOIN IDN_CLAIM_MAPPING AS MAPPING ON MAPPING.EXT_CLAIM_ID=CLAIMS.ID  AND CLAIMS.TENANT_ID=MAPPING.TENANT_ID INNER JOIN IDN_CLAIM AS RESOLVED          ON RESOLVED.ID=MAPPING.MAPPED_LOCAL_CLAIM_ID AND RESOLVED.TENANT_ID=MAPPING.TENANT_ID WHERE CLAIMS.DIALECT_ID=(SELECT ID FROM IDN_CLAIM_DIALECT WHERE DIALECT_URI=:1  AND TENANT_ID=:2 ) AND CLAIMS.TENANT_ID=:3 , OriginalSql = SELECT CLAIMS.ID, PROPERTY.PROPERTY_NAME, PROPERTY.PROPERTY_VALUE, CLAIMS.CLAIM_URI, RESOLVED.CLAIM_URI AS MAPPED_URI FROM IDN_CLAIM AS CLAIMS LEFT JOIN IDN_CLAIM_PROPERTY AS PROPERTY ON PROPERTY.LOCAL_CLAIM_ID=CLAIMS.ID AND CLAIMS.TENANT_ID=PROPERTY.TENANT_ID INNER JOIN IDN_CLAIM_MAPPING AS MAPPING ON MAPPING.EXT_CLAIM_ID=CLAIMS.ID  AND CLAIMS.TENANT_ID=MAPPING.TENANT_ID INNER JOIN IDN_CLAIM AS RESOLVED          ON RESOLVED.ID=MAPPING.MAPPED_LOCAL_CLAIM_ID AND RESOLVED.TENANT_ID=MAPPING.TENANT_ID WHERE CLAIMS.DIALECT_ID=(SELECT ID FROM IDN_CLAIM_DIALECT WHERE DIALECT_URI=? AND TENANT_ID=?) AND CLAIMS.TENANT_ID=?, Error Msg = ORA-00933: SQL command not properly ended```

**Comment by tharindu-bandara** [comment]
@dinusha92 could you please verify this scenario?

**Comment by dinusha92** [comment]
Hi @tharindu-bandara @tharakawijekoon this has happened due to the "AS" alias in table names. I will fix this and send the PR.

**Pull request merged**